File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/RequiredQuestionTest.java
package org.odk.collect.android.feature.formentry;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import org.apache.commons.csv.CSVRecord;
import org.junit.Ignore;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.RuleChain;
import org.junit.runner.RunWith;
import org.odk.collect.android.support.StorageUtils;
import org.odk.collect.android.support.rules.CollectTestRule;
import org.odk.collect.android.support.rules.TestRuleChain;

import java.io.IOException;
import java.util.List;

// Issue number NODK-249
@RunWith(AndroidJUnit4.class)
public class RequiredQuestionTest {

    public CollectTestRule rule = new CollectTestRule();

    @Rule
    public RuleChain copyFormChain = TestRuleChain.chain()
            .around(rule);

    @Test
    public void requiredQuestions_ShouldDisplayAsterisk_andCustomMessageIfSkipped() {
        rule.startAtMainMenu()
                .copyForm("requiredJR275.xml")
                .startBlankForm("required")
                .assertText("* Foo") //TestCase1
                .swipeToNextQuestionWithConstraintViolation("Custom required message");  //TestCase2
    }

    @Test
    public void pressingValidateInOptionsMenuOnSameScreen_shouldUseLatestAnswers() {
        rule.startAtMainMenu()
                .copyForm("requiredJR275.xml")
                .startBlankForm("required")
                .answerQuestion("Foo", true, "blah")
                .clickOptionsIcon()
                .clickOnString(org.odk.collect.strings.R.string.validate)
                .assertText(org.odk.collect.strings.R.string.success_form_validation)
                .assertTextDoesNotExist("Custom required message");
    }

    @Test
    public void pressingValidateInOptionsMenuOnDifferentScreen_shouldReturnToQuestionAndShowMessage() {
        rule.startAtMainMenu()
                .copyForm("requiredJR275.xml")
                .startBlankForm("required")
                .clickGoToArrow()
                .clickGoToEnd()
                .clickOptionsIcon()
                .clickOnString(org.odk.collect.strings.R.string.validate)
                .assertConstraintDisplayed("Custom required message")
                .assertQuestion("Foo", true);
    }

    @Test
    public void pressingValidateInOptionsMenuOnDifferentScreen_shouldReturnToQuestionAndShowMessage_whenTheQuestionIsInFieldList() {
        rule.startAtMainMenu()
                .copyForm("requiredQuestionInFieldList.xml")
                .startBlankForm("requiredQuestionInFieldList")
                .clickGoToArrow()
                .clickGoToEnd()
                .clickOptionsIcon()
                .clickOnString(org.odk.collect.strings.R.string.validate)
                .assertConstraintDisplayed("Custom required message")
                .assertQuestion("Foo", true);
    }

    @Test
    public void ifRequiredQuestionIsEmpty_shouldNotBeSavedToAuditLogWhenMovingForward() throws IOException {
        rule.startAtMainMenu()
                .copyForm("requiredJR275.xml")
                .startBlankForm("required")
                .swipeToNextQuestionWithConstraintViolation("Custom required message");

        List<CSVRecord> auditLog = StorageUtils.getAuditLogForFirstInstance();
        assertThat(auditLog.size(), equalTo(1));
        assertThat(auditLog.get(0).get(0), equalTo("form start"));
    }

    @Test
    public void ifRequiredQuestionIsEmpty_shouldNotBeSavedToAuditLogWhenFormValidated() throws IOException {
        rule.startAtMainMenu()
                .copyForm("requiredJR275.xml")
                .startBlankForm("required")
                .clickOptionsIcon()
                .clickOnString(org.odk.collect.strings.R.string.validate);

        List<CSVRecord> auditLog = StorageUtils.getAuditLogForFirstInstance();
        assertThat(auditLog.size(), equalTo(1));
        assertThat(auditLog.get(0).get(0), equalTo("form start"));
    }

    @Test
    @Ignore("https://github.com/getodk/collect/issues/5939")
    public void ifRequiredQuestionIsInFieldListAndNotFirst_shouldBeValidatedProperly() {
        rule.startAtMainMenu()
                .copyForm("requiredQuestionInFieldList.xml")
                .startBlankForm("requiredQuestionInFieldList")
                .answerQuestion(0, "Foo")
                .swipeToNextQuestionWithConstraintViolation("Custom required message2")
                .clickOptionsIcon()
                .clickOnString(org.odk.collect.strings.R.string.validate)
                .assertText("Custom required message2")
                .clickGoToArrow()
                .clickGoToEnd()
                .clickSaveAndExitWithError("Custom required message2");
    }

    @Test // https://github.com/getodk/collect/issues/5847
    public void savingFormWithInvalidQuestion_shouldNotChangeTheCurrentQuestionIndex() {
        rule.startAtMainMenu()
                .copyForm("two-question-required.xml")
                .startBlankForm("Two Question Required")
                .clickSave()
                .swipeToNextQuestion("What is your age?");
    }
}

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formentry/RequiredQuestionTest.kt
package org.odk.collect.android.feature.formentry

import androidx.test.ext.junit.runners.AndroidJUnit4
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.Rule
import org.junit.Test
import org.junit.rules.RuleChain
import org.junit.runner.RunWith
import org.odk.collect.android.support.StorageUtils.getAuditLogForFirstInstance
import org.odk.collect.android.support.rules.CollectTestRule
import org.odk.collect.android.support.rules.TestRuleChain.chain
import org.odk.collect.strings.R

@RunWith(AndroidJUnit4::class)
class RequiredQuestionTest {
    var rule: CollectTestRule = CollectTestRule()

    @get:Rule
    var ruleChain: RuleChain = chain()
        .around(rule)

    @Test
    fun requiredQuestionIsMarkedWithAnAsterisk() {
        rule.startAtMainMenu()
            .copyForm("required_question_with_custom_error_message.xml")
            .startBlankForm("required_question_with_custom_error_message")
            .assertQuestion("Required question", true)
    }

    @Test // https://github.com/getodk/collect/issues/6327
    fun requiredQuestionWithAudioIsMarkedWithAnAsterisk() {
        rule.startAtMainMenu()
            .copyForm("required_question_with_audio.xml")
            .startBlankForm("required_question_with_audio")
            .swipeToNextQuestion("Required question with audio", true)
    }

    @Test
    fun requiredQuestionDisplaysACustomErrorMessageIfSpecified() {
        rule.startAtMainMenu()
            .copyForm("required_question_with_custom_error_message.xml")
            .startBlankForm("required_question_with_custom_error_message")
            .swipeToNextQuestionWithConstraintViolation("Custom message")
    }

    @Test
    fun validatingFormByPressingValidateInOptionsMenuOnSameScreen_usesNewlyAddedAnswers() {
        rule.startAtMainMenu()
            .copyForm("required_question_with_custom_error_message.xml")
            .startBlankForm("required_question_with_custom_error_message")
            .answerQuestion("* Required question", "blah")
            .clickOptionsIcon()
            .clickOnString(R.string.validate)
            .assertText(R.string.success_form_validation)
            .assertTextDoesNotExist("Custom message")
    }

    @Test
    fun validatingFormByPressingValidateInOptionsMenuOnDifferentScreen_movesToTheQuestionWithErrorAndDisplaysError() {
        rule.startAtMainMenu()
            .copyForm("required_question_with_custom_error_message.xml")
            .startBlankForm("required_question_with_custom_error_message")
            .clickGoToArrow()
            .clickGoToEnd()
            .clickOptionsIcon()
            .clickOnString(R.string.validate)
            .assertConstraintDisplayed("Custom message")
            .assertQuestion("Required question", true)
    }

    @Test
    fun validatingFormByPressingValidateInOptionsMenuOnDifferentScreen_movesToTheQuestionWithErrorAndDisplaysError_whenTheQuestionIsInFieldList() {
        rule.startAtMainMenu()
            .copyForm("requiredQuestionInFieldList.xml")
            .startBlankForm("requiredQuestionInFieldList")
            .clickGoToArrow()
            .clickGoToEnd()
            .clickOptionsIcon()
            .clickOnString(R.string.validate)
            .assertConstraintDisplayed("Custom required message") // Make sure both questions are still displayed on the same screen
            .assertQuestion("Foo", true)
            .assertQuestion("Bar", true)
    }

    @Test
    fun emptyRequiredQuestion_isNotSavedToAuditLogOnMovingForward() {
        rule.startAtMainMenu()
            .copyForm("required_question_with_custom_error_message.xml")
            .startBlankForm("required_question_with_custom_error_message")
            .swipeToNextQuestionWithConstraintViolation("Custom message")

        val auditLog = getAuditLogForFirstInstance()
        assertThat(auditLog.size, equalTo(1))
        assertThat(auditLog[0][0], equalTo("form start"))
    }

    @Test
    fun emptyRequiredQuestion_isNotSavedToAuditLogOnFormValidation() {
        rule.startAtMainMenu()
            .copyForm("required_question_with_custom_error_message.xml")
            .startBlankForm("required_question_with_custom_error_message")
            .clickOptionsIcon()
            .clickOnString(R.string.validate)

        val auditLog = getAuditLogForFirstInstance()
        assertThat(auditLog.size, equalTo(1))
        assertThat(auditLog[0][0], equalTo("form start"))
    }

    @Test
    fun emptyRequiredQuestionInFieldListAndNotFirst_isValidatedProperly() {
        rule.startAtMainMenu()
            .copyForm("requiredQuestionInFieldList.xml")
            .startBlankForm("requiredQuestionInFieldList")
            .answerQuestion("Foo", true, "blah")
            .swipeToNextQuestionWithConstraintViolation("Custom required message2")
            .clickOptionsIcon()
            .clickOnString(R.string.validate)
            .assertText("Custom required message2")
            .clickGoToArrow()
            .clickGoToEnd()
            .clickSaveAndExitWithError("Custom required message2")
    }

    @Test // https://github.com/getodk/collect/issues/5847
    fun savingFormWithInvalidQuestion_doesNotChangeTheCurrentQuestionIndex() {
        rule.startAtMainMenu()
            .copyForm("two-question-required.xml")
            .startBlankForm("Two Question Required")
            .clickSave()
            .swipeToNextQuestion("What is your age?")
    }
}
